src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -1555,6 +1574,13 @@ AddInsertSelectCasts(List *insertTargetList, List *selectTargetList,
 		Form_pg_attribute attr = TupleDescAttr(destTupleDescriptor,
 											   insertEntry->resno - 1);
 
+        /*
+         * Assert that insert and select target entries created by
+         * ReorderInsertSelectTargetLists() have the expressions of the same
+         * type.
+         */
+        Assert(((Var *) insertEntry->expr)->vartype == exprType(selectEntry->expr));
+
 		Oid sourceType = insertColumn->vartype;
```

Comment:
@marcocitus, I realized that this is why we don't actually need to replace `insertColumn->vartype` with `exprType(selectEntry->expr)`.

We had to do that in our debugging session since we were not calling ReorderInsertSelectTargetLists() at that time.